### PR TITLE
feat(@aws-amplify/datastore): custom pk - customer facing types

### PR DESCRIPTION
**Pull request by manueliglesias**
#### Description of changes

Type definitions changes to support different kinds of identifiers:
- `ManagedIdentifier`
- `OptionallyManagedIdentifier`
- `CustomIdentifier`

#### Issue #, if available

N/A

#### Description of how you validated changes
The changes were validated by emulating the output of aws-amplify/amplify-codegen#351
See `packages/datastore/__tests__/custom-pk-typings.test.ts`

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/__tests__/custom-pk-typings.test.ts [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+});
+
+describe('CustomId - multi field', () => {
+	type PostCustomIdMetaData = {
+		identifier: CustomIdentifier<'tenantId' | 'postId'>;
+		readOnlyFields: 'createdAt' | 'updatedAt';
+	};
+	const PostCustomId = createDummyModelInstance(
+		class PostCustomId {
+			// readonly id: string;
```
Comment (david-mcafee) [inline packages/datastore/__tests__/custom-pk-typings.test.ts:157]: Remove?

File: packages/datastore/src/datastore/datastore.ts [hunk 2]
```diff
@@ -418,7 +423,7 @@ const createModelClass = <T extends PersistentModel>(
 							? <ModelInstanceMetadata>(<unknown>init)
 							: <ModelInstanceMetadata>{};
 
-					const { id: _id } = modelInstanceMetadata;
+					const { id: _id } = modelInstanceMetadata; // TODO: Here
```
Comment (david-mcafee) [inline packages/datastore/src/datastore/datastore.ts:426]: TODO?

File: packages/datastore/src/types.ts [hunk 3]
```diff
@@ -349,36 +347,85 @@ export type TypeConstructorMap = Record<
...
-> = DeepWritable<Omit<T, 'id' | K['readOnlyFields']>> &
-	Readonly<Pick<T, 'id' | K['readOnlyFields']>>;
+> = DeepWritable<
+	Omit<T, keyof IdentifierFields<M['identifier']> | M['readOnlyFields']>
+> &
+	Readonly<IdentifierFields<M['identifier']> & Pick<T, M['readOnlyFields']>>;
 
 export type ModelInstanceMetadata = {
-	id: string;
+	id: string; // TODO: Problematic!!!
```
Comment (david-mcafee) [inline packages/datastore/src/types.ts:428]: Perhaps explain more?

**Review by david-mcafee — COMMENTED** [review]
Left a few minor comments regarding commented code / TODOs.

File: packages/datastore/src/types.ts (hunk 3 above)
Comment (manueliglesias) [reply packages/datastore/src/types.ts:428]: Yeah, thanks for catching this. I left this as a reminder for me to mention this on the other big PR.

I'll remove the comment

**Review by iartemiev — APPROVED** [review]
LGTM 👍🏻

**Review by david-mcafee — APPROVED** [review]
LGTM! 🎉

**Pull request merged**